File: python/problems/leetcode/0139_word_break/solution.py
"""LeetCode 139: Word Break."""

from __future__ import annotations


class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> bool:
        words = set(wordDict)
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True
        max_len = max((len(w) for w in words), default=0)
        for i in range(1, n + 1):
            for j in range(max(0, i - max_len), i):
                if dp[j] and s[j:i] in words:
                    dp[i] = True
                    break
        return dp[n]


if __name__ == "__main__":
    assert Solution().wordBreak("leetcode", ["leet", "code"])
    assert Solution().wordBreak("applepenapple", ["apple", "pen"])
    assert not Solution().wordBreak("catsandog", ["cats", "dog", "sand", "and", "cat"])
    print("0139 OK")

File: cpp/problems/leetcode/0139_word_break/solution.cpp
// LeetCode 139
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> words(wordDict.begin(), wordDict.end());
        int n = (int)s.size();
        vector<bool> dp(n + 1, false);
        dp[0] = true;
        int max_len = 0;
        for (auto& w : words) max_len = max(max_len, (int)w.size());
        for (int i = 1; i <= n; ++i)
            for (int j = max(0, i - max_len); j < i; ++j)
                if (dp[j] && words.count(s.substr(j, i - j))) {
                    dp[i] = true;
                    break;
                }
        return dp[n];
    }
};

int main() {
    vector<string> d1{"leet", "code"};
    assert(Solution().wordBreak("leetcode", d1));
    vector<string> d2{"apple", "pen"};
    assert(Solution().wordBreak("applepenapple", d2));
    vector<string> d3{"cats", "dog", "sand", "and", "cat"};
    assert(!Solution().wordBreak("catsandog", d3));
    cout << "0139 OK" << endl;
    return 0;
}
